rust/lance/src/session/index_extension.rs

```diff
@@ -5,12 +5,34 @@ use std::sync::Arc;
 
 use lance_core::Result;
 use lance_file::reader::FileReader;
+use lance_index::{IndexParams, IndexType};
 
 use crate::{index::vector::VectorIndex, Dataset};
 
+pub trait IndexExtension: Send + Sync {
+    fn index_type(&self) -> IndexType;
+
+    fn to_scalar_index_extension(self: Arc<Self>) -> Option<Arc<dyn ScalarIndexExtension>>;
```

Comment:
`as_scalar() / as_vector()` does not create a new instance? Or add comments if my understanding is wrong.

Also this struct already has `IndexExtension` in the name.

Reply:
no new instance is created. The self capture is `self: Arc<Self>`. Each implementation will have it's own method to convert. This is kinda like a simple visitor pattern to recover types.

Reply:
I'll keep the `to_*` naming tho. since `as_*` usually implies returning a ref in rust